### PR TITLE
bpo-36025: Restore original function API for PyDate_FromTimestamp

### PR DESCRIPTION
**Pull request by pganssle**
Per pyo3/pyo3#352, I believe we've broken the C API for `PyDateTimeAPI->Date_FromTimestamp`. As a test (and to make sure we don't do it again), I've developed these tests. They are currently passing on 3.7 and they should be failing on master / Python 3.8 since we're passing a `tuple` to a function that strictly takes `int` or `float`.

Once I verify that these tests are indeed failing on CI, (they are anomalously passing in my local directory against master), I will also fix the underlying problem.

We don't need a news item for this, as it fixes a regression that was only ever released in alpha.

https://bugs.python.org/issue36025

### CODE REVIEW

**Comment by pganssle** [comment]
Ah, found the reason why it's anomalously passing - I accidentally shadowed the failing test with a passing one. 😅

File: Modules/_datetimemodule.c [hunk 1]
```diff
@@ -2905,7 +2905,7 @@ date_today(PyObject *cls, PyObject *dummy)
 @classmethod
 datetime.date.fromtimestamp
 
-    timestamp: object
+    timestamp: 'O'
```
Comment (serhiy-storchaka) [inline Modules/_datetimemodule.c]: Why this change?

File: Modules/_datetimemodule.c [hunk 2]
```diff
@@ -2916,11 +2916,23 @@ as local time.
...
 static PyObject *
 datetime_date_fromtimestamp(PyTypeObject *type, PyObject *timestamp)
-/*[clinic end generated code: output=fd045fda58168869 input=eabb3fe7f40491fe]*/
+/*[clinic end generated code: output=fd045fda58168869 input=64c24605543f2783]*/
 {
     return date_fromtimestamp((PyObject *) type, timestamp);
 }
 
+static PyObject *
+datetime_date_fromtimestamp_capi(PyObject *cls, PyObject *args) {
```
Comment (serhiy-storchaka) [inline Modules/_datetimemodule.c]: Move `{` on a new line for PEP 7.

File: Modules/_datetimemodule.c [hunk 3]
```diff
@@ -2916,11 +2916,23 @@ as local time.
...
+    PyObject *timestamp;
+    PyObject *result = NULL;
+
+    if (PyArg_ParseTuple(args, "O:fromtimestamp", &timestamp)) {
```
Comment (serhiy-storchaka) [inline Modules/_datetimemodule.c]: `PyArg_UnpackTuple()` is more efficient.

File: Modules/_testcapimodule.c [hunk 4]
```diff
@@ -2335,6 +2335,80 @@ get_timezone_utc_capi(PyObject* self, PyObject *args) {
     }
 }
 
+static PyObject *
+get_date_fromtimestamp(PyObject* self, PyObject *args) {
+    int ts;
+    int macro = 0;
+    if (!PyArg_ParseTuple(args, "i|p", &ts, &macro)) {
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: No need to convert timestamp to C long. Use "O" instead of "i". And you can use `PyTuple_Pack()` for creating an arguments tuple.

File: Modules/_testcapimodule.c [hunk 5]
```diff
@@ -2335,6 +2335,80 @@ get_timezone_utc_capi(PyObject* self, PyObject *args) {
     }
 }
 
+static PyObject *
+get_date_fromtimestamp(PyObject* self, PyObject *args) {
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: Move `{` on a new line for PEP 7.

File: Modules/_testcapimodule.c [hunk 6]
```diff
@@ -2335,6 +2335,80 @@ get_timezone_utc_capi(PyObject* self, PyObject *args) {
...
+    if ((pyts = PyLong_FromLong(ts)) == NULL) {
+        goto cleanup;
+    }
+
+    PyTuple_SET_ITEM(tsargs, 0, pyts);
+
+    // Pass along to the API function
+    if (macro) {
+        rv = PyDate_FromTimestamp(tsargs);
+    } else  {
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: Move `else` on a new line for PEP 7.

Remove duplicated space.

**Comment by serhiy-storchaka** [comment]
Since it fixes a regression in released version, I think it should be documented with a news entry.

**Comment by pganssle** [comment]
@serhiy-storchaka To be clear, this is only a regression on the 3.8 branch. I didn't think alpha versions counted as "released versions", but if they do I can add a news entry.

File: Modules/_datetimemodule.c (hunk 1 above)
Comment (pganssle) [reply Modules/_datetimemodule.c]: That was an accident, I was playing around with the argument clinic spec to try to get it to generate a `METH_VARARGS` method. Obviously it didn't work, so I'll revert it.

**Comment by pganssle** [comment]
@serhiy-storchaka Thanks for the review, I made all the suggested changes, except I haven't added a NEWS entry yet.

File: Modules/_testcapimodule.c [hunk 7]
```diff
@@ -2335,6 +2335,71 @@ get_timezone_utc_capi(PyObject* self, PyObject *args) {
...
+{
+    int macro = 0;
+    int usetz = 0;
+    PyObject *tsargs = NULL, *ts = NULL, *tzinfo = Py_None, *rv = NULL;
+    if (!PyArg_ParseTuple(args, "OO|pp", &ts, &tzinfo, &usetz, &macro)) {
+        return NULL;
+    }
+
+    // Construct the argument tuple
+    if ((tsargs = PyTuple_New(1 + usetz)) == NULL) {
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: Would not be simpler to use `PyTuple_Pack()` here too?
```c
if (usetz) {
    tsargs = PyTuple_Pack(2, ts, tzinfo);
}
else {
    tsargs = PyTuple_Pack(1, ts);
}
```

File: Modules/_testcapimodule.c [hunk 8]
```diff
@@ -4732,9 +4797,11 @@ static PyMethodDef TestMethods[] = {
     {"datetime_check_tzinfo",     datetime_check_tzinfo,         METH_VARARGS},
     {"make_timezones_capi",     make_timezones_capi,             METH_NOARGS},
     {"get_timezones_offset_zero",   get_timezones_offset_zero,   METH_NOARGS},
-    {"get_timezone_utc_capi",    get_timezone_utc_capi,            METH_VARARGS},
-    {"test_list_api",           test_list_api,                   METH_NOARGS},
-    {"test_dict_iteration",     test_dict_iteration,             METH_NOARGS},
+    {"get_timezone_utc_capi",    get_timezone_utc_capi,          METH_VARARGS},
+    {"get_date_fromtimestamp",   get_date_fromtimestamp,         METH_VARARGS},
+    {"get_datetime_fromtimestamp", get_datetime_fromtimestamp,   METH_VARARGS},
+    {"test_list_api",           (PyCFunction)test_list_api,      METH_NOARGS},
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: I don't think an explicit cast to `PyCFunction` is needed. Both functions have compatible signatures.

File: Modules/_testcapimodule.c (hunk 7 above)
Comment (pganssle) [reply Modules/_testcapimodule.c]: Hm, yeah, that should work. Good call.

File: Modules/_testcapimodule.c (hunk 8 above)
Comment (pganssle) [reply Modules/_testcapimodule.c]: I didn't add this, it got pulled into the diff because of formatting changes I think. Either way, I'll remove it.

Comment (pganssle) [reply Modules/_testcapimodule.c]: Oh actually I think this was because this branch was originally created on Python 3.7 so I would know that the tests would have passed at the time. I think the function casts were dropped between then and now, but I pulled them in because of a merge conflict.

Fixed now.

**Comment by pganssle** [comment]
@serhiy-storchaka Does this look good? Can we merge or are we waiting on something else?

**Comment by pganssle** [comment]
CC @abalkin

**Review by abalkin — APPROVED** [review]

**Comment by pganssle** [comment]
The first force push since approval was an accident - included a What's New entry that was intended for another PR. The second one corrects that mistake and rebases this against the current master.

@abalkin @ambv Can we merge this before the next alpha release? It would be good to get at least the pyo3 3.8-dev build job working again.

**Review by serhiy-storchaka — APPROVED** [review]

File: Modules/_testcapimodule.c [hunk 9]
```diff
@@ -2335,6 +2335,73 @@ get_timezone_utc_capi(PyObject* self, PyObject *args) {
...
+{
+    PyObject *tsargs = NULL, *ts = NULL, *rv = NULL;
+    int macro = 0;
+
+    if (!PyArg_ParseTuple(args, "O|p", &ts, &macro)) {
+        return NULL;
+    }
+
+    // Construct the argument tuple
+    if((tsargs = PyTuple_Pack(1, ts)) == NULL) {
```
Comment (ZackerySpytz) [inline Modules/_testcapimodule.c]:
```suggestion
    if ((tsargs = PyTuple_Pack(1, ts)) == NULL) {
```

File: Modules/_testcapimodule.c [hunk 10]
```diff
@@ -2335,6 +2335,73 @@ get_timezone_utc_capi(PyObject* self, PyObject *args) {
...
+{
+    int macro = 0;
+    int usetz = 0;
+    PyObject *tsargs = NULL, *ts = NULL, *tzinfo = Py_None, *rv = NULL;
+    if (!PyArg_ParseTuple(args, "OO|pp", &ts, &tzinfo, &usetz, &macro)) {
+        return NULL;
+    }
+
+    // Construct the argument tuple
+    Py_INCREF(ts);
```
Comment (ZackerySpytz) [inline Modules/_testcapimodule.c]: There's no need to incref `ts` (`PyTuple_Pack()` will take care of that).

File: Modules/_testcapimodule.c [hunk 11]
```diff
@@ -2335,6 +2335,73 @@ get_timezone_utc_capi(PyObject* self, PyObject *args) {
...
+    if (usetz) {
+        Py_INCREF(tzinfo);
```
Comment (ZackerySpytz) [inline Modules/_testcapimodule.c]: The same goes for `tzinfo`.

**Review by ZackerySpytz — COMMENTED** [review]
There are reference leaks.

**Comment by pganssle** [comment]
@ZackerySpytz Thanks for the review! You are right about the refcounts, I have removed them. I guess `PyTuple_Pack` "steals" the existing reference to both its arguments, so I don't have to increase or decrease the reference count, just pass it into the tuple. I've checked that the reference count is unchanged before and after calling `get_datetime_fromtimestamp` and it seems it is now!

**Comment by pganssle** [comment]
@abalkin @serhiy-storchaka Next alpha release is coming up soon. Anything else you'd like me to do before merge?

**Review by berkerpeksag — APPROVED** [review]
